### PR TITLE
Introducing epoch manager

### PR DESCRIPTION
**Pull request by yontyon**
In this PR we introduce the epoch manager and demonstrate its power via add/remove node command with lagged replicas.
**Epoch** is a life cycle of the system. by changing epoch we indicate that we actually started a new system on top of an old state.
All replicas have to decide on starting a new epoch exactly at the same state to preserve consistency. Thus, starting a new epoch after wedge makes great sense.
Practically, when a replica decides to start a new epoch, it writes a new "genesis" block that contains this new epoch under a pre-known key.

### CODE REVIEW

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 1]
```diff
@@ -346,6 +341,14 @@ void BCStateTran::init(uint64_t maxNumOfRequiredStoredCheckpoints,
 
       ConcordAssertEQ(getFetchingState(), FetchingState::NotFetching);
     }
+    {
```
Comment (toly-kournik) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:344]: All the initialization was a single transaction.

File: bftengine/include/bftengine/EpochManager.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#include <functional>
+
+namespace bftEngine {
+class EpochManager : public bftEngine::ResPagesClient<EpochManager, 1> {
+  struct EpochData : public concord::serialize::SerializableFactory<EpochData> {
+    uint64_t epochNumber_;
+    EpochData() = default;
+    EpochData(uint64_t epochNumber) : epochNumber_{epochNumber} {}
+
+    const std::string getVersion() const override { return "1"; }
```
Comment (toly-kournik) [inline bftengine/include/bftengine/EpochManager.hpp:27]: no need in `getVersion()` anymore

File: bftengine/include/bftengine/EpochManager.hpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    page.resize(sizeOfReservedPage());
+    metrics_.Register();
+  }
+
+ public:
+  static EpochManager& instance() {
+    static EpochManager instance_;
+    return instance_;
+  }
+  uint64_t getEpochNumber(bool fromPages) {
```
Comment (toly-kournik) [inline bftengine/include/bftengine/EpochManager.hpp:45]: what's use case when you don't want to get the most updated version?

File: bftengine/include/bftengine/EpochManager.hpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    if (!fromPages) return epochNumber_;
+    if (!loadReservedPage(0, sizeOfReservedPage(), page.data())) return 0;
+    EpochData edata;
+    std::istringstream inStream;
+    inStream.str(page);
+    concord::serialize::Serializable::deserialize(inStream, edata);
+    epochNumber_ = edata.epochNumber_;
+    return edata.epochNumber_;
+  }
+  void setEpochNumber(uint64_t newEopch, bool toPages) {
```
Comment (toly-kournik) [inline bftengine/include/bftengine/EpochManager.hpp:55]: Same here, why not to always persist epoch num?

File: bftengine/include/bftengine/EpochManager.hpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    std::ostringstream outStream;
+    concord::serialize::Serializable::serialize(outStream, edata);
+    auto data = outStream.str();
+    saveReservedPage(0, data.size(), data.data());
+  }
+
+  void startNewEpoch() { startNewEpoch_ = true; }
+  void markEpochAsStarted() { startNewEpoch_ = false; }
+  bool isNewEpoch() { return startNewEpoch_; }
+  void setNewEpochFlagHandler(const std::function<void(bool)>& cb) { setNewEpochFlagCallBack_ = cb; }
```
Comment (toly-kournik) [inline bftengine/include/bftengine/EpochManager.hpp:70]: consider using callback subscription mechanism as in ST callbacks if suitable

File: bftengine/include/bftengine/EpochManager.hpp (hunk 3 above)
Comment (yontyon) [reply bftengine/include/bftengine/EpochManager.hpp:45]: in some cases, the replica wants to get its own epoch number and not the one that was updated in the reserved pages from other replicas.
actually, the only reason to use reserved pages is for read only replicas.

File: bftengine/include/bftengine/EpochManager.hpp (hunk 4 above)
Comment (yontyon) [reply bftengine/include/bftengine/EpochManager.hpp:55]: we want to persist the epoch number only when we sure that other replicas do that again. If we catch up from a previous epoch, we want to update only our local epoch and not in the reserved pages

File: bftengine/include/bftengine/EpochManager.hpp (hunk 5 above)
Comment (yontyon) [reply bftengine/include/bftengine/EpochManager.hpp:70]: I have only a single callback here, if in the future there will be more, I will change it

File: bftengine/include/bftengine/EpochManager.hpp (hunk 3 above)
Comment (toly-kournik) [reply bftengine/include/bftengine/EpochManager.hpp:45]: it complicates the logic, hard to understand

File: bftengine/include/bftengine/EpochManager.hpp (hunk 5 above)
Comment (toly-kournik) [reply bftengine/include/bftengine/EpochManager.hpp:70]: why to use a flag?

File: kvbc/src/Replica.cpp [hunk 6]
```diff
@@ -158,6 +159,40 @@ void Replica::createReplicaAndSyncState() {
       std::terminate();
     }
   }
+
+  uint64_t epoch = 0;
```
Comment (toly-kournik) [inline kvbc/src/Replica.cpp:163]: `createReplicaAndSyncState()` becomes a beast, maybe make a separate function?

File: bftengine/include/bftengine/EpochManager.hpp (hunk 5 above)
Comment (toly-kournik) [reply bftengine/include/bftengine/EpochManager.hpp:70]: `setNewEpochHandler()` ?

Comment (yontyon) [reply bftengine/include/bftengine/EpochManager.hpp:68]: we sometimes (after state transfer) want to cancel the decision to start a new epoch

**Review by toly-kournik — APPROVED** [review]
LGTM!

**Pull request merged**